test/sm4test.c

```diff
@@ -0,0 +1,72 @@
+/*
+ * Copyright 2017 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright 2017 [Ribose Inc.](https://www.ribose.com). All Rights Reserved.
+ *
+ * Licensed under the OpenSSL license (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include "testutil.h"
+
+#ifndef OPENSSL_NO_SM4
+# include "internal/sm4.h"
+
+static int test_sm4_ecb(void)
+{
+    static const uint8_t k[16] = {
+       0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef,
+       0xfe, 0xdc, 0xba, 0x98, 0x76, 0x54, 0x32, 0x10
+    };
+
+    static const uint8_t input[16] = {
+       0x01, 0x23, 0x45, 0x67, 0x89, 0xab, 0xcd, 0xef,
+       0xfe, 0xdc, 0xba, 0x98, 0x76, 0x54, 0x32, 0x10
+    };
+
+    static const uint8_t expected[16] = {
+       0x68, 0x1e, 0xdf, 0x34, 0xd2, 0x06, 0x96, 0x5e,
+       0x86, 0xb3, 0xe9, 0x4f, 0x53, 0x6e, 0x42, 0x46
+    };
+
+    // after 1,000,000 iterations
+    static const uint8_t expected_iter[16] = {
+       0x59, 0x52, 0x98, 0xc7, 0xc6, 0xfd, 0x27, 0x1f,
+       0x04, 0x02, 0xf8, 0x04, 0xc3, 0x3d, 0x3f, 0x66
+    };
+
+    SM4_KEY key;
+    SM4_set_key(k, &key);
+
+    uint8_t block[16];
```

Comment:
Can't use SM4_BLOCK_SIZE here - but perhaps a define instead of a magic number here (and other instances below) would be better?

Reply:
Actually, since `internal/sm4.h` is included, it's perfectly possible to use `SM4_BLOCK_SIZE`...

Reply:
Using `SM4_BLOCK_SIZE` now, moved to `test/sm4_internal_test.c`.